File: 00-reference/node-console/00-reference.js
#!/usr/bin/env node
/** Console grid navigator: username login and 3×3 keyboard navigation. */

const readline = require("readline");

const ROWS = ["t", "m", "b"];
const COLS = ["l", "m", "r"];

function formatPos(row, col) {
  return `${ROWS[row]}/${COLS[col]}`;
}

function tryMove(row, col, dr, dc) {
  const nr = Math.max(0, Math.min(2, row + dr));
  const nc = Math.max(0, Math.min(2, col + dc));
  if (nr === row && nc === col) return { row, col, moved: false };
  return { row: nr, col: nc, moved: true };
}

function drawCell(selected) {
  if (selected) {
    return ["┏━━━┓", "┃ X ┃", "┗━━━┛"];
  }
  return ["┌───┐", "│   │", "└───┘"];
}

function render(username, row, col, previous) {
  console.clear();
  const prevText = previous ? formatPos(previous.row, previous.col) : "—";
  console.log(`Name: ${username}`);
  console.log(`Current position: ${formatPos(row, col)}`);
  console.log(`Previous position: ${prevText}`);
  console.log("\nUse arrow keys or WASD to move (q to quit).\n");

  // Draw one grid row at a time: 3 cells wide, 3 terminal lines tall per row.
  for (let r = 0; r < 3; r++) {
    const topLine = [];
    const midLine = [];
    const botLine = [];
    for (let c = 0; c < 3; c++) {
      const lines = drawCell(r === row && c === col);
      topLine.push(lines[0]);
      midLine.push(lines[1]);
      botLine.push(lines[2]);
    }
    console.log(topLine.join(" "));
    console.log(midLine.join(" "));
    console.log(botLine.join(" "));
  }
}

function askUsername() {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });
  return new Promise((resolve) => {
    const prompt = () => {
      rl.question("Username: ", (answer) => {
        const name = answer.trim();
        if (!name) {
          console.log("Username is required.");
          prompt();
          return;
        }
        rl.close();
        resolve(name);
      });
    };
    console.log("Login\n");
    prompt();
  });
}

async function runGrid(username) {
  let row = 1;
  let col = 1;
  let previous = null;

  readline.emitKeypressEvents(process.stdin);
  if (process.stdin.isTTY) process.stdin.setRawMode(true);
  process.stdin.resume();

  render(username, row, col, previous);

  process.stdin.on("keypress", (str, key) => {
    if (key.ctrl && key.name === "c") {
      process.stdin.setRawMode(false);
      process.stdin.pause();
      process.exit(0);
    }
    let dir = null;
    if (key.name === "up" || str === "w" || str === "W") dir = "up";
    else if (key.name === "down" || str === "s" || str === "S") dir = "down";
    else if (key.name === "left" || str === "a" || str === "A") dir = "left";
    else if (key.name === "right" || str === "d" || str === "D") dir = "right";
    else if (str === "q" || str === "Q") dir = "quit";
    if (dir === "quit") {
      process.stdin.setRawMode(false);
      process.stdin.pause();
      process.exit(0);
    }
    if (!dir) return;

    let dr = 0;
    let dc = 0;
    if (dir === "up") dr = -1;
    else if (dir === "down") dr = 1;
    else if (dir === "left") dc = -1;
    else if (dir === "right") dc = 1;

    const result = tryMove(row, col, dr, dc);
    if (result.moved) {
      previous = { row, col };
      row = result.row;
      col = result.col;
    }
    render(username, row, col, previous);
  });
}

askUsername().then(runGrid);

File: 00-reference/python-console/00-reference.py
#!/usr/bin/env python3
"""Console grid navigator: username login and 3×3 keyboard navigation."""

import curses

ROWS = ("t", "m", "b")
COLS = ("l", "m", "r")


def format_pos(row: int, col: int) -> str:
    return f"{ROWS[row]}/{COLS[col]}"


def try_move(row: int, col: int, dr: int, dc: int) -> tuple[int, int, bool]:
    """Return new (row, col) and whether the position changed."""
    new_row = max(0, min(2, row + dr))
    new_col = max(0, min(2, col + dc))
    if new_row == row and new_col == col:
        return row, col, False
    return new_row, new_col, True


def read_username(stdscr: curses.window) -> str:
    """Prompt for a non-empty username (no password)."""
    curses.echo()
    stdscr.clear()
    stdscr.addstr(0, 0, "Login")
    stdscr.addstr(2, 0, "Username: ")
    stdscr.refresh()
    while True:
        username = stdscr.getstr(2, 10, 40).decode("utf-8").strip()
        if username:
            curses.noecho()
            return username
        stdscr.addstr(4, 0, "Username is required. Try again.")
        stdscr.clrtoeol()
        stdscr.refresh()


def draw_cell(stdscr: curses.window, y: int, x: int, selected: bool) -> None:
    """Draw one grid cell; selected cells show X with no color highlight."""
    top = "┏━━━┓" if selected else "┌───┐"
    mid = "┃ X ┃" if selected else "│   │"
    bot = "┗━━━┛" if selected else "└───┘"
    stdscr.addstr(y, x, top)
    stdscr.addstr(y + 1, x, mid)
    stdscr.addstr(y + 2, x, bot)


def draw_screen(
    stdscr: curses.window,
    username: str,
    row: int,
    col: int,
    previous: tuple[int, int] | None,
) -> None:
    stdscr.clear()
    prev_text = format_pos(*previous) if previous else "—"
    stdscr.addstr(0, 0, f"Name: {username}")
    stdscr.addstr(1, 0, f"Current position: {format_pos(row, col)}")
    stdscr.addstr(2, 0, f"Previous position: {prev_text}")
    stdscr.addstr(4, 0, "Use arrow keys or WASD to move (L to logout, Q to quit).")

    base_y, base_x = 6, 2
    cell_w = 6
    for r in range(3):
        for c in range(3):
            draw_cell(
                stdscr,
                base_y + r * 4,
                base_x + c * cell_w,
                r == row and c == col,
            )
    stdscr.refresh()


def run_grid(stdscr: curses.window, username: str) -> str:
    row, col = 1, 1
    previous: tuple[int, int] | None = None

    while True:
        draw_screen(stdscr, username, row, col, previous)
        key = stdscr.getch()

        if key in (ord("q"), ord("Q")):
            return "quit"
        if key in (ord("l"), ord("L")):
            return "logout"

        dr = dc = 0
        if key in (curses.KEY_UP, ord("w"), ord("W")):
            dr = -1
        elif key in (curses.KEY_DOWN, ord("s"), ord("S")):
            dr = 1
        elif key in (curses.KEY_LEFT, ord("a"), ord("A")):
            dc = -1
        elif key in (curses.KEY_RIGHT, ord("d"), ord("D")):
            dc = 1
        else:
            continue

        new_row, new_col, moved = try_move(row, col, dr, dc)
        if moved:
            previous = (row, col)
            row, col = new_row, new_col


def main(stdscr: curses.window) -> None:
    curses.curs_set(0)
    stdscr.keypad(True)

    while True:
        username = read_username(stdscr)
        if run_grid(stdscr, username) == "quit":
            break


if __name__ == "__main__":
    curses.wrapper(main)
